app/src/main/java/com/eveningoutpost/dexdrip/Models/Bubble.java

```diff
@@ -0,0 +1,290 @@
+                int index = toInt(temp);
+                sb.append(snParts[index]);
+
+            }
+        } catch (Exception e) {
+            e.printStackTrace();
+            return "";
+        }
+        return "0" + sb.toString();
+    }
+
+    public static int toInt(String bi) {
+        int len = bi.length();
+        int sum = 0;
+        int tmp, max = len - 1;
+        for (int i = 0; i < len; ++i) {
+            tmp = bi.charAt(i) - '0';
+            sum += tmp * Math.pow(2, max--);
+        }
+        return sum;
+    }
+
+    static int errorCount = 0;
+
+    public static BridgeResponse decodeBubblePacket(byte[] buffer, int len) {
+        final BridgeResponse reply = new BridgeResponse();
+        int count = Pref.getInt("Cobblecount", 0);
+//        Log.e(TAG, "decodeBubblePacket" + Arrays.toString(buffer));
+        if (buffer[0] == -128) {
+            PersistentStore.setString("Bubblebattery", Integer.toString(buffer[4]));
+            Pref.setInt("bridge_battery", buffer[4]);
+//            String data=DexCollectionService.bytesToHexString(buffer);
+            PersistentStore.setString("BubbleHArdware", Integer.toString(buffer[2]) + ".0");
+            PersistentStore.setString("BubbleFirmware", Integer.toString(buffer[1]) + ".0");
+//            PersistentStore.setString("LibreSN", SensorSn);
+            ByteBuffer ackMessage = ByteBuffer.allocate(6);
+            ackMessage.put(0, (byte) 0x02);
+            ackMessage.put(1, (byte) 0x01);
+            ackMessage.put(2, (byte) 0x00);
+            ackMessage.put(3, (byte) 0x00);
+            ackMessage.put(4, (byte) 0x00);
+            ackMessage.put(5, (byte) 0x2B);
+            reply.add(ackMessage);
+            bytes = null;
+            dateAll = false;
+            s_full_data = null;
+            Log.e(TAG, "reset data");
+            return getBubbleResponse();
+        }
+        if (buffer[0] == -64) {
+            int i2 = 0;
+            ArrayList<Byte> bytett = new ArrayList();
+            for (byte b : buffer) {
+                if (i2 >= 2) {
+                    bytett.add(b);
+                }
+                i2++;
+            }
+            Byte[] byte2 = new Byte[4];
+            String UUID = HexDump.dumpHexString(bytett.subList(0, 8).toArray(byte2));
+            String SensorSn = serNumber(UUID);
+            if (!SensorSn.equals(PersistentStore.getString("LibreSN"))) {
+                Sensor.stopSensor();
+                Sensor.create(System.currentTimeMillis());
+            }
+            PersistentStore.setString("LibreSN", SensorSn);
+            return reply;
+        }
+        if (buffer[0] == -126) {
+            int expectedSize = lens;
+
+            if (s_full_data == null) {
+                InitBuffer(expectedSize);
+            }
+            if (bytes == null) {
+                bytes = new ArrayList();
+            }
+            addData(buffer);
+            PersistentStore.setBoolean("CobbleNoSensor", false);
```

Comment:
To aid refactoring and reduce chances for human error with typos, preference key strings should be stored as static final strings and referenced from those strings rather than with repeated string literals.